src/jit/jithashtable.h

```diff
@@ -0,0 +1,672 @@
+                if (m_node != nullptr)
+                {
+                    return;
+                }
+
+                // Otherwise...
+                m_index++;
+            }
+            while ((m_index < m_tableSize) && (m_table[m_index] == nullptr))
+            {
+                m_index++;
+            }
+
+            if (m_index >= m_tableSize)
+            {
+                m_node = nullptr;
+                return;
+            }
+            else
+            {
+                m_node = m_table[m_index];
+            }
+            assert(m_node != nullptr);
+        }
+
+        bool Equal(const KeyIterator& i) const
+        {
+            return i.m_node == m_node;
+        }
+
+        void operator++()
+        {
+            Next();
+        }
+
+        void operator++(int)
+        {
+            Next();
+        }
+    };
+
+    Value& operator[](Key k) const
+    {
+        Value* p = LookupPointer(k);
+        assert(p);
+        return *p;
+    }
+
+private:
+    // Find the next prime number >= the given value.
+    static JitPrimeInfo NextPrime(unsigned number)
+    {
+        for (int i = 0; i < (int)(sizeof(jitPrimeInfo) / sizeof(jitPrimeInfo[0])); i++)
+        {
+            if (jitPrimeInfo[i].prime >= number)
+            {
+                return jitPrimeInfo[i];
+            }
+        }
+
+        // overflow
+        Behavior::NoMemory();
+    }
+
+    // The node type.
+    struct Node
+    {
+        Node* m_next; // Assume that the alignment requirement of Key and Value are no greater than Node*, so put m_next
+                      // to avoid unnecessary padding.
+        Key   m_key;
+        Value m_val;
+
+        Node(Key k, Value v, Node* next) : m_next(next), m_key(k), m_val(v)
+        {
+        }
+
+        void* operator new(size_t sz, Allocator* alloc)
+        {
+            return alloc->Alloc(sz);
+        }
+
+        void operator delete(void* p, Allocator* alloc)
+        {
+            alloc->Free(p);
+        }
+    };
+
+    // Instance members
+    Allocator*   m_alloc;         // Allocator to use in this table.
+    Node**       m_table;         // pointer to table
+    JitPrimeInfo m_tableSizeInfo; // size of table (a prime) and information about it
+    unsigned     m_tableCount;    // number of elements in table
+    unsigned     m_tableMax;      // maximum occupied count
+};
+
+// A few simple KeyFuncs types...
+
+// Base class for types whose equality function is the same as their "==".
+template <typename T>
+struct JitKeyFuncsDefEquals
+{
+    static bool Equals(const T& x, const T& y)
+    {
+        return x == y;
+    }
+};
+
+template <typename T>
+struct JitPtrKeyFuncs : public JitKeyFuncsDefEquals<const T*>
+{
+public:
+    static unsigned GetHashCode(const T* ptr)
+    {
+        // Hmm.  Maybe (unsigned) ought to be "ssize_t" -- or this ought to be ifdef'd by size.
+        return static_cast<unsigned>(reinterpret_cast<uintptr_t>(ptr));
```

Comment:
I don't claim to be an expert in these matters, but my guess is that using the lower 32-bits is probably a safe bet, given that it is likely that different pointers to the same type are unlikely to differ in the higher bits.

Reply:
Yes, and the use of `ssize_t` isn't really an option, the hash code needs to stay as 32 bit otherwise we need to make `JitPrimeInfo` use 64 bit as well and that would make hashtable objects larger.

All this came from `SimplerHashTable`. I was careful not to change anything beyond `IAllocator` related code because it's more difficult to evaluate changes to hashtables and other collections. But I'll fix the comments :)